016-database-adapters.md

```diff
@@ -0,0 +1,340 @@
+  [Optimize default DB indices #506](https://github.com/appwrite/appwrite/issues/506)
+  [Add Fields param in ListDocument and GetDocument APIs #499](https://github.com/appwrite/appwrite/issues/499)
+  [MongoDB in Appwrite #909](https://github.com/appwrite/appwrite/issues/909)
+
+## Summary
+
+[summary]: #summary
+
+<!-- Brief explanation of the proposed contribution. Write your answer below. -->
+
+Appwrite's database doesn't have a built-in way to customize DB indexes or under the hood adapters. The idea in this RFC is to make both options available directly from the database dashboard.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+<!-- Write your answer below. -->
+
+The current implementation of the Appwrite database is very good and quick to get started easily, but doesn't allow enough flexibility for when your data set grows or when you need to implement more advanced use-cases.
+
+**What is the context or background in which this problem exists?**
+
+Larger data set and more complex projects, require more flexibility and transparency regarding the database inner structure and performance capabilities.
+
+**Once the proposal is implemented, how will the system change?**
+
+We will implement simplified data structures that could be manipulated from both the Appwrite dashboard and directly from the relevant connected DB. We will also add new endpoints to easily add, delete, and view collection rules and indexes.
+
+<!-- Write your answer below. -->
+
+<!-- Please avoid discussing your proposed solution. -->
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail keeping in mind the following:
+
+- Its interaction with other parts of the system is clear
+- It is reasonably clear how the contribution would be implemented
+- Dependencies on libraries, tools, projects or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**
+
+Write your answer below.
+
+-->
+
+### Utopia Database
+
+We'll implement an extra database abstraction layer as part of the Utopia project. This layer will include support for multiple database, and will expose a consistent API which will act as the lowest common denominator for all the supported adapters.
+
+The list of supported adapters will include:
+* Postgres
+* MySQL
+* MariaDB
+* MongoDB¹
```

Comment:
When it comes time to write documentation for these changes, I think it would be really nice to have a section outlining the differences between the supported databases. 

Although I know by design they will all function the same, there might be subtle differences that developers might want to know.

Developers using Appwrite will most likely care about:
- Disk size of database
- Portability, (ease of backup/restore)
- Speed
- Resource cost (CPU, memory, etc)

Reply:
Just a comment for your consideration. Are you considering aggregation functions? Such as count, sum, max, min, etc. of a particular rule in the collection, this could be quite useful.

https://en.m.wikipedia.org/wiki/Aggregate_function

Reply:
Probably not in step one, but I can definitely see the value here as we build the future roadmap of this API